pkg/authorization/authorizer/authorizer.go

```diff
@@ -171,63 +177,61 @@ func (a *openshiftAuthorizer) Authorize(passedAttributes AuthorizationAttributes
 	}
 
-	globalAuthorizationResult, globalReason, err := a.authorizeWithNamespaceRules(a.masterAuthorizationNamespace, attributes)
-	if err != nil {
-		return false, "", err
-	}
-	switch globalAuthorizationResult {
-	case Allow:
+	// keep track of errors in case we are unable to authorize the action.
+	// It is entirely possible to get an error and be able to continue determine authorization status in spite of it.
+	// This is most common when a bound role is missing, but enough roles are still present and bound to authorize the request.
+	errs := []error{}
+
+	globalAllowed, globalReason, err := a.authorizeWithNamespaceRules(a.masterAuthorizationNamespace, attributes)
+	if globalAllowed {
 		return true, globalReason, nil
-	case Deny:
-		return false, globalReason, nil
+	}
+	if err != nil {
+		errs = append(errs, err)
 	}
 
 	if len(attributes.GetNamespace()) != 0 {
-		namespaceAuthorizationResult, namespaceReason, err := a.authorizeWithNamespaceRules(attributes.GetNamespace(), attributes)
-		if err != nil {
-			return false, "", err
-		}
-		switch namespaceAuthorizationResult {
-		case Allow:
+		namespaceAllowed, namespaceReason, err := a.authorizeWithNamespaceRules(attributes.GetNamespace(), attributes)
+		if namespaceAllowed {
 			return true, namespaceReason, nil
-		case Deny:
-			return false, namespaceReason, nil
 		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return false, "", kerrors.NewAggregate(errs)
 	}
 
 	return false, "denied by default", nil
 }
 
-type authorizationResult string
-
-const (
-	Allow   = authorizationResult("allow")
-	Deny    = authorizationResult("deny")
-	Unknown = authorizationResult("unknown")
-)
-
-func (a *openshiftAuthorizer) authorizeWithNamespaceRules(namespace string, passedAttributes AuthorizationAttributes) (authorizationResult, string, error) {
+// authorizeWithNamespaceRules returns isAllowed, reason, and error.  If an error is returned, isAllowed and reason are still valid.  This seems strange
+// but errors are not always fatal to the authorization process.  It is entirely possible to get an error and be able to continue determine authorization
+// status in spite of it.  This is most common when a bound role is missing, but enough roles are still present and bound to authorize the request.
+func (a *openshiftAuthorizer) authorizeWithNamespaceRules(namespace string, passedAttributes AuthorizationAttributes) (bool, string, error) {
 	attributes, ok := passedAttributes.(openshiftAuthorizationAttributes)
 	if !ok {
-		return Deny, "", fmt.Errorf("attributes are not of expected type: %#v", attributes)
+		return false, "", fmt.Errorf("attributes are not of expected type: %#v", attributes)
 	}
 
-	allRules, err := a.getEffectivePolicyRules(namespace, attributes.GetUserInfo())
-	if err != nil {
-		return Deny, "", err
+	allRules, ruleRetrievalError := a.getEffectivePolicyRules(namespace, attributes.GetUserInfo())
+	if allRules == nil {
```

Comment:
since the `nil` case and the empty case both result in returning `false, "", ruleRetrievalError`, can we skip the nil check here?